File: 05-BruteForce/Eugene/1520_내리막길.py
m,n = map(int, input().split())
maps = [list(map(int, input().split())) for _ in range(m)]
visited = [[-1] * n for _ in range(m)]    # 방문 여부 확인

def dfs(x, y):
    cnt = 0
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]

    if x == m-1 and y == n-1:             # 끝까지 도달하는 경우
        return 1
    
    if visited[x][y] != -1:               # 이미 더 낮다는걸 확인 한 경우
        return visited[x][y]

    for i in range(4):                     # 4방향 탐색
        nx = x + dx[i]
        ny = y + dy[i]
        
        if 0 <= nx < m and 0 <= ny < n:    # 맵 안에 있는 경우  
            if maps[x][y] > maps[nx][ny]:  # 갈 수 있는 경우
                cnt += dfs(nx, ny)                # 죽지 않고 돌아온 재귀함수
                
    visited[x][y] = cnt
    return visited[x][y]
    
print(dfs(0,0))

File: 06-DP/haeun/1520_내리막길.cpp
#include <iostream> 
#include <vector> 
#include <string>
#include <algorithm>
#include <queue> 
using namespace std;

const int MAX = 501;
int height[MAX][MAX]; 
int dp[MAX][MAX];

int m, n;
int cnt = 0;
int dx[] = {0, 0, -1, 1}; 
int dy[] = {-1, 1, 0, 0}; 

void input() {
	cin >> m >> n; 

	for(int i = 0; i < m; i++){
		for(int j = 0; j < n; j++){
			cin >> height[i][j]; 
			dp[i][j] = -1; 
		}
	}
}

int dfs(int x, int y){ 
	if(x == m - 1 && y == n - 1) return 1; // --- (a) 
	if(dp[x][y] != -1) return dp[x][y]; // --- (b) 

	dp[x][y] = 0;
	for(int i = 0; i < 4; i++){ 
		int nx = x + dx[i]; 
		int ny = y + dy[i]; 

		if(nx < 0 || nx >= m || ny < 0 || ny >= n) continue; 

		if(height[x][y] > height[nx][ny]){
			dp[x][y] += dfs(nx, ny); 
		}
	}

	return dp[x][y]; // --- (c) 
}

int main() {
	ios::sync_with_stdio(0);
    cin.tie(0);

	input();

	cout << dfs(0, 0); 
	
	return 0;
}